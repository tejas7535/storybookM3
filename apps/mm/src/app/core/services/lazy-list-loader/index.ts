export * from './lazy-list-loader.service';
