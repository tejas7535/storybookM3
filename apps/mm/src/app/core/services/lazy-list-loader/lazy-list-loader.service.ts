import { Injectable } from '@angular/core';

import { map, Observable } from 'rxjs';

import { ListValue } from '@mm/shared/models/lazy-list-loader/mm-list-value.model';

import { environment } from '../../../../environments/environment';
import {
  MMComplexResponse,
  MMResponseVariants,
  MMSimpleResponse,
} from '../../../shared/models';
import { RestService } from '../';

const isComplex = (
  response: MMResponseVariants
): response is MMComplexResponse => !Array.isArray(response.data);

const isSimple = (response: MMResponseVariants): response is MMSimpleResponse =>
  !isComplex(response) && 'data' in response.data[0];

@Injectable({
  providedIn: 'root',
})
export class LazyListLoaderService {
  public baseImageURL = environment.baseUrl.replace('/v1', '');

  public constructor(private readonly restService: RestService) {}

  public loadOptions(
    url: string,
    values: { name: string; value: string | number }[]
  ): Observable<ListValue[]> {
    // eslint-disable-next-line unicorn/no-array-reduce
    const requestUrl = values.reduce(
      (filledUrl, { name, value }) =>
        filledUrl.replace(`<${name}>`, `${value}`),
      url
    );

    return this.restService.getLoadOptions(requestUrl).pipe(
      map((response) => {
        if (isComplex(response)) {
          return response.data.bearingSeats.map(
            ({ _media, data: { id, title } }) => ({
              id,
              title,
              image: this.getImageUrl(_media[0].href),
            })
          );
        }
        if (isSimple(response)) {
          return response.data.map(({ _media, data }) => ({
            image: _media ? this.getImageUrl(_media[0]?.href) : undefined,
            ...data,
          }));
        }

        return response.data;
      }),
      map((items) =>
        items.map(({ id, title, image }) => ({
          id,
          text: title,
          imageUrl: image ?? undefined,
        }))
      )
    );
  }

  private getImageUrl(wrongImageUrl: string): string {
    const fileName = this.extractFilename(wrongImageUrl);

    return `${this.baseImageURL}/Images/${fileName}`;
  }

  private extractFilename(url: string): string {
    const filename = url.slice(Math.max(0, url.lastIndexOf('/') + 1));

    return filename;
  }
}
