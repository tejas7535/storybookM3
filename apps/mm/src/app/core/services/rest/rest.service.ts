import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { map, Observable } from 'rxjs';

import { CalculationRequestPayload } from '@mm/shared/models/calculation-request.model';
import { withCache } from '@ngneat/cashew';

import { environment } from '../../../../environments/environment';
import {
  MMBearingPreflightResponse,
  PreflightRequestBody,
  SearchResult,
  ShaftMaterialResponse,
} from '../../../shared/models';
import { BearinxOnlineResult } from '../bearinx-result.interface';

@Injectable({
  providedIn: 'root',
})
export class RestService {
  private currentLanguage: string;

  private readonly bearingCalculationPath = `${environment.baseUrl}/calculate`;

  public constructor(private readonly httpClient: HttpClient) {}

  public setCurrentLanguage(language: string): void {
    this.currentLanguage = language;
  }

  public getBearingSearch(searchQuery: string): Observable<SearchResult> {
    return this.httpClient.get<SearchResult>(
      `${environment.baseUrl}/bearings/search?pattern=${searchQuery}&page=1&size=1000`
    );
  }

  public getBearingCalculationResult(
    requestPayload: CalculationRequestPayload
  ): Observable<BearinxOnlineResult> {
    return this.httpClient.post<BearinxOnlineResult>(
      this.bearingCalculationPath,
      requestPayload,
      {
        context: withCache({ version: JSON.stringify(requestPayload) }),
      }
    );
  }

  public getBearingPreflightResponse(
    body: PreflightRequestBody
  ): Observable<MMBearingPreflightResponse> {
    const path = `${environment.baseUrl}/dialog`;

    return this.httpClient.post<MMBearingPreflightResponse>(path, body);
  }

  public getBearingsMaterialResponse(
    idmmShaftMaterial: string
  ): Observable<ShaftMaterialResponse> {
    return this.httpClient.get<ShaftMaterialResponse>(
      `${environment.baseUrl}/materials/${idmmShaftMaterial}`,
      { context: withCache({ version: this.currentLanguage }) }
    );
  }

  public getLoadOptions<T>(requestUrl: string): Observable<T> {
    return this.httpClient.get<T>(requestUrl, {
      context: withCache({ version: this.currentLanguage }),
    });
  }

  public getBearinxVersions(): Observable<{ [key: string]: string }> {
    return this.httpClient
      .get<
        { name: string; version: string }[]
      >(`${environment.bearinxApiBaseUrl}/version`)
      .pipe(
        map((response) =>
          Object.fromEntries(
            response.map(({ name, version }) => [name, version])
          )
        )
      );
  }
}
