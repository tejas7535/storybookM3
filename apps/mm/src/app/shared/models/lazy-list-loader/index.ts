export * from './mm-bearings-material-response.model';
export * from './mm-bearings-preflight-response.model';
export * from './mm-response-variants.model';
