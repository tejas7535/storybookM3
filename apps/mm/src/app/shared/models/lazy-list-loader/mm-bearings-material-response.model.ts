export interface MMBearingsMaterialResponse {
  id: string;
  IDMM_MODULUS_OF_ELASTICITY: string;
  IDMM_POISSON_RATIO: string;
}
