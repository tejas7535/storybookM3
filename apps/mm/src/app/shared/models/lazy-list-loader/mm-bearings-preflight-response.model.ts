export interface MMBearingPreflightField {
  id: string;
  range: { id: string; title: string }[] | null;
  defaultValue: string;
}

export interface MMBearingPreflightResponse {
  data: {
    input: {
      id: string;
      title: string;
      fields: MMBearingPreflightField[];
    }[];
  };
}
