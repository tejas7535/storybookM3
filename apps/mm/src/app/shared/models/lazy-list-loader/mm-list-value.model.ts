export interface ListValue {
  id: string;
  text: string | null;
  imageUrl?: string;
}
