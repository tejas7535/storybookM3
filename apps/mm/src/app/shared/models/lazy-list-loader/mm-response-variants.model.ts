export interface MMResponse {
  id: string;
  title: string;
  image?: string | null;
}

export interface MMBaseResponse {
  data: MMResponse[];
}

export interface MMSimpleResponse {
  data: { data: MMResponse; _media?: [{ href: string }] }[];
}

export interface MMComplexResponse {
  data: {
    bearingSeats: {
      data: {
        id: string;
        title: string;
      };
      _media: [{ href: string }];
    }[];
  };
}

export type MMResponseVariants =
  | MMBaseResponse
  | MMSimpleResponse
  | MMComplexResponse;
