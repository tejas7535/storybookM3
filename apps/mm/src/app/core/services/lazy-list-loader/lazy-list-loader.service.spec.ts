import { of } from 'rxjs';

import {
  BEARING_SEATS_RESPONSE,
  SIMPLE_LIST_RESPONSE,
} from '@mm/testing/mocks/rest.service.mock';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { RestService } from '../rest';
import { LazyListLoaderService } from './lazy-list-loader.service';

jest.mock('@mm/environments/environment', () => ({
  environment: {
    baseUrl: 'http://mock-base-url/mounting/api/v2/mountingmanager',
  },
}));

describe('LazyListLoaderService', () => {
  let service: LazyListLoaderService;
  let spectator: SpectatorService<LazyListLoaderService>;
  let restService: RestService;

  const createService = createServiceFactory({
    service: LazyListLoaderService,
    providers: [
      {
        provide: RestService,
        useValue: { getLoadOptions: jest.fn() },
      },
    ],
  });

  beforeEach(() => {
    spectator = createService();
    service = spectator.service;
    restService = spectator.inject(RestService);
  });

  describe('#loadBearingSeatsOptions', () => {
    it('should call loadBearingSeatsOptions url', (done) => {
      jest
        .spyOn(restService, 'getLoadOptions')
        .mockReturnValue(of(BEARING_SEATS_RESPONSE));
      const mockUrl = `bearingSeatsURL`;

      service.loadBearingSeatsOptions(mockUrl).subscribe((response) => {
        expect(response).toMatchSnapshot();
        done();
      });

      expect(service['restService'].getLoadOptions).toHaveBeenCalledWith(
        mockUrl
      );
    });
  });

  describe('#loadOptions', () => {
    it('should call getLoadOptions simple request', (done) => {
      const mockUrl = `simple`;

      jest
        .spyOn(restService, 'getLoadOptions')
        .mockReturnValue(of(SIMPLE_LIST_RESPONSE));

      service.loadOptions(mockUrl).subscribe((response) => {
        expect(response).toMatchSnapshot();
        done();
      });

      expect(service['restService'].getLoadOptions).toHaveBeenCalledWith(
        mockUrl
      );
    });
  });
});
