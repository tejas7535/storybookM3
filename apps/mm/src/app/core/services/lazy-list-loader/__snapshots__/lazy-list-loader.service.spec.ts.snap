// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyListLoaderService #loadBearingSeatsOptions should call loadBearingSeatsOptions url 1`] = `
Array [
  Object {
    "id": "mockId",
    "imageUrl": "http://mock-base-url/mounting/api/images/imageName.png",
    "text": "mockTitle",
  },
]
`;

exports[`LazyListLoaderService #loadOptions should call getLoadOptions simple request 1`] = `
Array [
  Object {
    "id": "mockId",
    "imageUrl": "http://mock-base-url/mounting/api/images/imageName1.png",
    "text": "mockTitle",
  },
  Object {
    "id": "mockId2",
    "imageUrl": undefined,
    "text": "mockTitle2",
  },
]
`;
